[FIRRTL] ModuleInliner: Add support for RefType

prithayan opened the pull request:
This PR adds support to handle inlining of RefType ops. Temporary wires of
RefType that are created to replace ports of inlined modules, must be removed.
As a post-processing step, the pass iterates over all the `RefType` wires, which
were added during the inlining transformation. Then replaces all the uses of the
wire with its driver, and marks the wire for removal. Some of the uses of the
wire need to be moved to ensure the driver dominates the uses.

seldridge commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -729,6 +729,28 @@ void Inliner::cloneAndRename(
     newAnnotations.push_back(anno);
   }
 
+  // Handle inlining of RefType ops. Temporary wires of RefType must be removed
+  // here. The RefType wires were created to replace ports of inlined modules.
+  // RefType ports can only have out Direction. And there is only one connect to
+  // the output reftype port. Given these constraints, remove the connect, and
+  // directly replace the users with the connect.getSrc(). To ensure
+  // hierarchical inlining is correctly done, update any temporary RefType wires
+  // from the mapper.
```
We're relaxing the downwards only restriction meaning that we can now see `RefType` ports with input direction.

Both the comment and the logic should be updated to handle this.

seldridge reviewed (commented):
I like that this is as minimal of a PR as it is.

We should refactor this pass (in the future) to clean up the creation of the wire. This is tricky to untangle, though.

Main question is making sure this works for the new upwards reads.

prithayan commented:
@seldridge , updated this to run as a post processing step. Now it iterates over all RefType wires, and replaces the users with the driver. This also has to move the users after the definition, to work for the "u-turn" of RefType ports kind of use cases.  Also updated the lit tests.
